contracts/pallet-ibc/src/ics20_fee/mod.rs

```diff
@@ -0,0 +1,281 @@
+		channel_id: &ChannelId,
+		counterparty: &Counterparty,
+		version: &Version,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_open_init(
+			&mut ctx,
+			output,
+			order,
+			connection_hops,
+			port_id,
+			channel_id,
+			counterparty,
+			version,
+			relayer,
+		)
+	}
+
+	fn on_chan_open_try(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		order: Order,
+		connection_hops: &[ConnectionId],
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		counterparty: &Counterparty,
+		version: &Version,
+		counterparty_version: &Version,
+		relayer: &Signer,
+	) -> Result<Version, Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_open_try(
+			&mut ctx,
+			output,
+			order,
+			connection_hops,
+			port_id,
+			channel_id,
+			counterparty,
+			version,
+			counterparty_version,
+			relayer,
+		)
+	}
+
+	fn on_chan_open_ack(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		counterparty_version: &Version,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_open_ack(
+			&mut ctx,
+			output,
+			port_id,
+			channel_id,
+			counterparty_version,
+			relayer,
+		)
+	}
+
+	fn on_chan_open_confirm(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_open_confirm(&mut ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_chan_close_init(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_close_init(&mut ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_chan_close_confirm(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_close_confirm(&mut ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_recv_packet(
+		&self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		packet: &Packet,
+		relayer: &Signer,
+	) -> Result<Acknowledgement, Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		let packet_data =
+			serde_json::from_slice::<PacketData>(packet.data.as_slice()).map_err(|e| {
+				Ics04Error::implementation_specific(format!("Failed to decode packet data {:?}", e))
+			})?;
+		let percent = ServiceCharge::<T>::get().unwrap_or(T::ServiceCharge::get());
+		let fee = percent * packet_data.token.amount.as_u256().as_u128();
```

Comment:
Are we sure this doesn't truncate?

Reply:
We ensure the amount token amount is never above the max value a u128 allows in the ics20 callback, since account balances are u128